File: sdk/java/src/main/java/com/anyvali/interchange/Exporter.java
package com.anyvali.interchange;

import com.anyvali.AnyValiDocument;
import com.anyvali.ExportMode;
import com.anyvali.Schema;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Export schemas to AnyVali interchange format.
 */
public final class Exporter {
    private Exporter() {}

    /**
     * Export a schema to an AnyVali document map.
     */
    public static Map<String, Object> exportSchema(Schema<?> schema) {
        return exportSchema(schema, ExportMode.PORTABLE, null, null);
    }

    /**
     * Export a schema to an AnyVali document map.
     */
    public static Map<String, Object> exportSchema(Schema<?> schema, ExportMode mode,
                                                    Map<String, ? extends Schema<?>> definitions,
                                                    Map<String, Object> extensions) {
        Map<String, Object> rootNode = schema.toPortableNode();

        Map<String, Object> defs = new LinkedHashMap<>();
        if (definitions != null) {
            for (var entry : definitions.entrySet()) {
                defs.put(entry.getKey(), entry.getValue().toPortableNode());
            }
        }

        Map<String, Object> ext = new LinkedHashMap<>();
        if (mode == ExportMode.EXTENDED && extensions != null) {
            ext.putAll(extensions);
        }

        AnyValiDocument doc = new AnyValiDocument(rootNode, defs, ext);
        return doc.toMap();
    }

    /**
     * Export a schema to a JSON string.
     */
    public static String exportSchemaJson(Schema<?> schema) {
        return exportSchemaJson(schema, ExportMode.PORTABLE, null, null);
    }

    /**
     * Export a schema to a JSON string.
     */
    public static String exportSchemaJson(Schema<?> schema, ExportMode mode,
                                          Map<String, ? extends Schema<?>> definitions,
                                          Map<String, Object> extensions) {
        Map<String, Object> doc = exportSchema(schema, mode, definitions, extensions);
        return JsonHelper.toJson(doc);
    }
}

File: sdk/kotlin/src/main/kotlin/com/anyvali/interchange/Exporter.kt
package com.anyvali.interchange

import com.anyvali.*
import com.anyvali.schemas.*
import kotlinx.serialization.json.*

object Exporter {

    fun exportToJson(schema: Schema, mode: ExportMode = ExportMode.PORTABLE): String {
        val doc = schema.export(mode)
        return toJsonString(doc)
    }

    fun toJsonString(doc: AnyValiDocument): String {
        val json = buildJsonObject {
            put("anyvaliVersion", JsonPrimitive(doc.anyvaliVersion))
            put("schemaVersion", JsonPrimitive(doc.schemaVersion))
            put("root", doc.root)
            put("definitions", JsonObject(doc.definitions.mapValues { it.value }))
            put("extensions", JsonObject(doc.extensions))
        }
        return Json { prettyPrint = true }.encodeToString(JsonElement.serializer(), json)
    }

    fun exportDocument(
        schema: Schema,
        definitions: Map<String, Schema> = emptyMap(),
        extensions: Map<String, JsonElement> = emptyMap(),
        mode: ExportMode = ExportMode.PORTABLE
    ): AnyValiDocument {
        if (mode == ExportMode.PORTABLE && schema.hasCustomValidators) {
            throw ValidationError(
                listOf(
                    ValidationIssue(
                        code = IssueCodes.CUSTOM_VALIDATION_NOT_PORTABLE,
                        message = "Schema uses custom validators that are not portable"
                    )
                )
            )
        }
        return AnyValiDocument(
            root = schema.exportNode(),
            definitions = definitions.mapValues { it.value.exportNode() },
            extensions = extensions
        )
    }
}
